compiler/rustc_infer/src/infer/relate/equate.rs

```diff
@@ -159,8 +167,22 @@ impl<'tcx> TypeRelation<'tcx> for Equate<'_, '_, 'tcx> {
         }
 
         if a.skip_binder().has_escaping_bound_vars() || b.skip_binder().has_escaping_bound_vars() {
-            self.fields.higher_ranked_sub(a, b, self.a_is_expected)?;
-            self.fields.higher_ranked_sub(b, a, self.a_is_expected)?;
+            // Check whether `for<..> A == for<..> B` holds.
+            // For this to hold, we need to check `for<..> A <: for<..> B` and
+            // `for<..> B <: for<..> A` in an invariant context.
```

Comment:
these comments are wrong, `<:` means "is subtype of" and we are not subtyping in an invariant context. also affects a few lines further down

please add something like this
```rust
// When equating binders, we check that there is a 1-to-1
// correspondence between the bound vars in both types.
//
// We do so by separately instantiating one of the binders with
// placeholders and the other with inference variables and then
// equating the instantiated types. 
```